### PR TITLE
Add a custom evaluator option for RunApp

### LINKED ISSUE #22946: Add ability to use custom evaluation functions to RunApp
## Reason
Tests can arise where more granularity in evaluating the output is needed than can be provided by regex alone. 

## Design
Ideally some sort of solution would exist that allows us to add a custom script for evaluating a test's output

## Impact
Adds more functionality to RunApp in the TestHarness

### PR DESCRIPTION
**Pull request by socratesgorilla**
Closes #22946

### CODE REVIEW

File: modules/doc/content/python/TestHarness.md [hunk 1]
```diff
@@ -100,6 +100,33 @@ that MOOSE [uses](/application_usage/input_syntax.html optional=True). An exampl
 
 !listing moose/test/tests/kernels/simple_diffusion/tests
 
+## Test Evaluation and Custom Evaluators
+
+MOOSE has various setups for evaluating whether a created test is successful or not. Depending on what type of Tester you specify in the "test specification" file, these can range from comparing the output of JSON, XML, CSV, or Exodus files, or matching a pattern in the output of the test using a regular expression.
+
+Sometimes, it may not be possible to properly evaluate a test with the built-in checks MOOSE provides. In this situation, MOOSE has functionality for evaluating a test using a custom, user-supplied evaluation function. 
+
+To do this, create a Python script in the folder containing your test. The script should be blank aside from a single function named `custom_evaluation(output)`:
```
Comment (cticenhour) [inline modules/doc/content/python/TestHarness.md:109]:
```suggestion
To do this, create a Python script in the folder containing your test. The script should be blank aside from a single function named `custom_evaluation(output)`:

```

File: modules/doc/content/python/TestHarness.md [hunk 2]
```diff
@@ -100,6 +100,33 @@ that MOOSE [uses](/application_usage/input_syntax.html optional=True). An exampl
...
+```
+def custom_evaluation(output):
+  #Do your evaluation logic here
+  if output == "foo":
+    return True
+  return False
+```
+`output` is a string containing the test's entire output and can be manipulated as one sees fit. The function should return a boolean representing whether the test passed or not.
```
Comment (cticenhour) [inline modules/doc/content/python/TestHarness.md:117]:
```suggestion

`output` is a string containing the test's entire output and can be manipulated as one sees fit. The function should return a boolean representing whether the test passed or not.
```

File: modules/doc/content/python/TestHarness.md [hunk 3]
```diff
@@ -100,6 +100,33 @@ that MOOSE [uses](/application_usage/input_syntax.html optional=True). An exampl
...
+
+After creating the .py file, add `custom_evaluation_script = '[filename].py'` to your test specification file. For example:
```
Comment (cticenhour) [inline modules/doc/content/python/TestHarness.md:119]:
```suggestion
After creating the .py file, add `custom_evaluation_script = '[filename].py'` to your test specification file. For example:

```

File: modules/doc/content/python/TestHarness.md [hunk 4]
```diff
@@ -100,6 +100,33 @@ that MOOSE [uses](/application_usage/input_syntax.html optional=True). An exampl
...
+```
+[Tests]
+  [my_test]
+    type = RunApp
+    input = my_input.i
+    custom_evaluation_script = 'my_custom_eval.py'
+  [../]
```
Comment (cticenhour) [inline modules/doc/content/python/TestHarness.md:126]:
```suggestion
  []
```

**Review by cticenhour — CHANGES_REQUESTED** [review]
"MooseDown" Is a bit picky about needing empty lines between certain bits of content. This should fix your issue.

**Comment by cticenhour** [comment]
@socratesgorilla FYI - MALAMUTE failure in `Internal App Tests` is unrelated and due to a patch for another PR. Disregard.

File: test/tests/test_harness/custom_eval [hunk 5]
```diff
@@ -0,0 +1,7 @@
+[Tests]
```
Comment (cticenhour) [inline test/tests/test_harness/custom_eval:1]: Please also test the cases where the user might supply `absent_out` or `expect_out` and an error is generated.

File: python/TestHarness/tests/test_CustomEval.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+
+from TestHarnessTestCase import TestHarnessTestCase
+import subprocess
+class TestHarnessTester(TestHarnessTestCase):
+  def testCustomEval(self):
```
Comment (cticenhour) [inline python/TestHarness/tests/test_CustomEval.py:5]: See other comment - expand this testing to hit more lines of the additional python.

**Review by cticenhour — CHANGES_REQUESTED** [review]
This is neat! A few things to address to wrap this up.

**Review by cticenhour — COMMENTED** [review]
Found one more thing - the doco page builds, but doesn't look right: 

image

Stand by for another code adjustment once I get the site built.

File: modules/doc/content/python/TestHarness.md [hunk 7]
```diff
@@ -100,6 +100,36 @@ that MOOSE [uses](/application_usage/input_syntax.html optional=True). An exampl
...
+
+```
+[Tests]
+  [my_test]
+    type = RunApp
+    input = my_input.i
+    custom_evaluation_script = 'my_custom_eval.py'
+  []
+[]
+```
```
Comment (cticenhour) [inline modules/doc/content/python/TestHarness.md:131]: Gah - spaces! This line has an extra space at the end of it that needs to be removed to render the section below it properly. My setup of VSCode removes trailing whitespaces by default, so I recommend checking that option (or the equivalent in your editor) to catch things like this.

**Review by cticenhour — CHANGES_REQUESTED** [review]

**Comment by socratesgorilla** [comment]
@cticenhour failing tests seem to be unrelated to me -- everything look good now?

**Review by cticenhour — APPROVED** [review]
CRANE failure is unrelated, and Griffin failures are related to the move to the new GitHub server. Pending another failure, I think this is good to go once testing completes.

**Pull request merged**